🌠Use stdlib for trigonometric functions

sukibaby opened the pull request:
We can get a nice overall performance boost by replacing some clunky code in RageMath with standard library functions.

I don't know, maybe something was different in 2016, but I'm not seeing how the RageFast___ code was faster (https://github.com/itgmania/itgmania/commit/86470b797783ea95e16a578589158a1b90e27221)

In fact, I attempted to implement a fast non-standard library method of taking the sine but could not come close to the speed of `std::sin`, even though it was faster than Kyz's method.

**This even includes purposely slowing down `std::sin` to ensure it was measurable!**

Example output of my [RageFastSin test case:](https://cpp.sh/?source=%23include+%3Ciostream%3E%0D%0A%23include+%3Ccmath%3E%0D%0A%23include+%3Cchrono%3E%0D%0A%0D%0A%23ifndef+PI%0D%0A%23define+PI+3.14159265358979323846%0D%0A%23endif%0D%0A%0D%0A%2F%2F+KYZ%27S+RAGEFASTSIN%0D%0A%23define+SCALE(x%2C+l1%2C+h1%2C+l2%2C+h2)+(((x)+-+(l1))+*+((h2)+-+(l2))+%2F+((h1)+-+(l1))+%2B+(l2))%0D%0Atemplate%3Ctypename+T%2C+typename+U%3E%0D%0Ainline+U+lerp(T+x%2C+U+l%2C+U+h)%0D%0A%7B%0D%0A++++return+U(x+*+(h+-+l)+%2B+l)%3B%0D%0A%7D%0D%0Astatic+const+unsigned+int+sine_table_size+%3D+1024%3B%0D%0Astatic+const+unsigned+int+sine_index_mod+%3D+sine_table_size+*+2%3B%0D%0Astatic+const+double+sine_table_index_mult+%3D+static_cast%3Cdouble%3E(sine_index_mod)+%2F+(PI+*+2)%3B%0D%0Astatic+float+sine_table%5Bsine_table_size%5D%3B%0D%0Astruct+SineTableInitializer+%7B%0D%0A++++SineTableInitializer()+%7B%0D%0A++++++++for+(unsigned+int+i+%3D+0%3B+i+%3C+sine_table_size%3B+%2B%2Bi)+%7B%0D%0A++++++++++++float+angle+%3D+SCALE(i%2C+0%2C+sine_table_size%2C+0.0f%2C+PI)%3B%0D%0A++++++++++++sine_table%5Bi%5D+%3D+std%3A%3Asin(angle)%3B%0D%0A++++++++%7D%0D%0A++++%7D%0D%0A%7D%3B%0D%0Astatic+SineTableInitializer+sineTableInitializer%3B%0D%0Afloat+RageFastSinKyz(float+angle)%0D%0A%7B%0D%0A++++if+(angle+%3D%3D+0)+%7B+return+0%3B+%7D%0D%0A++++float+index+%3D+angle+*+sine_table_index_mult%3B%0D%0A++++int+first_index+%3D+static_cast%3Cint%3E(index)%3B%0D%0A++++int+second_index+%3D+(first_index+%2B+1)+%25+sine_index_mod%3B%0D%0A++++float+remainder+%3D+index+-+first_index%3B%0D%0A++++first_index+%25%3D+sine_index_mod%3B%0D%0A++++float+first+%3D+0.0f%3B%0D%0A++++float+second+%3D+0.0f%3B%0D%0A%23define+SET_SAMPLE(sample)+%5C%0D%0A++++if(sample%23%23_index+%3E%3D+static_cast%3Cint%3E(sine_table_size))+%5C%0D%0A++++%7B+%5C%0D%0A++++++++sample%3D+-sine_table%5Bsample%23%23_index+-+sine_table_size%5D%3B+%5C%0D%0A++++%7D+%5C%0D%0A++++else+%5C%0D%0A++++%7B+%5C%0D%0A++++++++sample%3D+sine_table%5Bsample%23%23_index%5D%3B+%5C%0D%0A++++%7D%0D%0A++++SET_SAMPLE(first)%3B%0D%0A++++SET_SAMPLE(second)%3B%0D%0A%23undef+SET_SAMPLE%0D%0A++++return+lerp(remainder%2C+first%2C+second)%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F+SUKIBABY%27S+RAGEFASTSIN%0D%0Afloat+RageFastSinSuki(float+angle)%0D%0A%7B%0D%0A++++angle+%3D+fmod(angle%2C+2+*+M_PI)%3B%0D%0A++++if+(angle+%3C+0)+angle+%2B%3D+2+*+M_PI%3B%0D%0A++++if+(angle+%3E+M_PI)+angle+-%3D+2+*+M_PI%3B%0D%0A++++const+float+B+%3D+4+%2F+M_PI%3B%0D%0A++++const+float+C+%3D+-4+%2F+(M_PI+*+M_PI)%3B%0D%0A++++float+y+%3D+B+*+angle+%2B+C+*+angle+*+std%3A%3Aabs(angle)%3B%0D%0A++++return+y%3B%0D%0A%7D%0D%0A%0D%0Aint+main()+%7B%0D%0A++++const+long+long+numIterations+%3D+5%3B%0D%0A++++float+angle+%3D+(0.69+*+2.0f+*+PI)%3B%0D%0A%0D%0A++++auto+startRageFastSinKyz+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A++++for+(long+long+i+%3D+0%3B+i+%3C+numIterations%3B+%2B%2Bi)+%7B%0D%0A++++++++RageFastSinKyz(angle)%3B%0D%0A++++%7D%0D%0A++++auto+endRageFastSinKyz+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A%0D%0A++++auto+startRageFastSinSuki+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A++++for+(long+long+i+%3D+0%3B+i+%3C+numIterations%3B+%2B%2Bi)+%7B%0D%0A++++++++RageFastSinSuki(angle)%3B%0D%0A++++%7D%0D%0A++++auto+endRageFastSinSuki+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A%0D%0A++++auto+startStdSin+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A++++volatile+float+accumulator+%3D+0%3B+%2F%2F+Use+volatile+to+prevent+optimization%0D%0A++++for+(long+long+i+%3D+0%3B+i+%3C+numIterations%3B+%2B%2Bi)+%7B%0D%0A++++++++accumulator+%2B%3D+std%3A%3Asin(angle+%2B+i)%3B+%2F%2F+Slightly+vary+the+input+to+prevent+caching%0D%0A++++%7D%0D%0A++++auto+endStdSin+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A++++%2F%2F+Use+the+accumulator+in+some+way+to+ensure+the+loop+isn%27t+optimized+out%0D%0A++++std%3A%3Acout+%3C%3C+%22Accumulator+(to+prevent+optimization)%3A+%22+%3C%3C+accumulator+%3C%3C+std%3A%3Aendl%3B%0D%0A%0D%0A++++auto+durationRageFastSinKyz+%3D+std%3A%3Achrono%3A%3Aduration_cast%3Cstd%3A%3Achrono%3A%3Amicroseconds%3E(endRageFastSinKyz+-+startRageFastSinKyz).count()%3B%0D%0A++++auto+durationRageFastSinSuki+%3D+std%3A%3Achrono%3A%3Aduration_cast%3Cstd%3A%3Achrono%3A%3Amicroseconds%3E(endRageFastSinSuki+-+startRageFastSinSuki).count()%3B%0D%0A++++auto+durationStdSin+%3D+std%3A%3Achrono%3A%3Aduration_cast%3Cstd%3A%3Achrono%3A%3Amicroseconds%3E(endStdSin+-+startStdSin).count()%3B%0D%0A%0D%0A++++std%3A%3Acout+%3C%3C+%22RageFastSin+(Kyz)+took+%22+%3C%3C+durationRageFastSinKyz+%3C%3C+%22+microseconds.%5Cn%22%3B%0D%0A++++std%3A%3Acout+%3C%3C+%22RageFastSin+(Sukibaby)+took+%22+%3C%3C+durationRageFastSinSuki+%3C%3C+%22+microseconds.%5Cn%22%3B%0D%0A++++std%3A%3Acout+%3C%3C+%22std%3A%3Asin+took+%22+%3C%3C+durationStdSin+%3C%3C+%22+microseconds.%5Cn%22%3B%0D%0A%0D%0A++++return+0%3B%0D%0A%7D%0D%0A)
```
Accumulator (to prevent optimization): 0.0651476
RageFastSin (Kyz) took 95 microseconds.
RageFastSin (Sukibaby) took 60 microseconds.
std::sin took 19 microseconds.
```

Example output of my [RageFastCos test case](https://cpp.sh/?source=%23include+%3Ciostream%3E%0D%0A%23include+%3Cchrono%3E%0D%0A%23include+%3Ccmath%3E%0D%0A%0D%0A%23define+SCALE(x%2C+l1%2C+h1%2C+l2%2C+h2)+(((x)+-+(l1))+*+((h2)+-+(l2))+%2F+((h1)+-+(l1))+%2B+(l2))%0D%0A%23define+PI+3.141592653589793238462643383279%0D%0A%0D%0Astatic+const+unsigned+int+sine_table_size+%3D+1024%3B%0D%0Astatic+const+unsigned+int+sine_index_mod+%3D+sine_table_size+*+2%3B%0D%0Astatic+const+double+sine_table_index_mult+%3D+static_cast%3Cdouble%3E(sine_index_mod)+%2F+(PI+*+2)%3B%0D%0Astatic+float+sine_table%5Bsine_table_size%5D%3B%0D%0A%0D%0A%2F%2F+Linear+interpolation+function%0D%0Afloat+lerp(float+t%2C+float+a%2C+float+b)+%7B%0D%0A++++return+a+%2B+t+*+(b+-+a)%3B%0D%0A%7D%0D%0A%0D%0Astruct+sine_initter+%7B%0D%0A++++sine_initter()+%7B%0D%0A++++++++for+(unsigned+int+i+%3D+0%3B+i+%3C+sine_table_size%3B+%2B%2Bi)+%7B%0D%0A++++++++++++float+angle+%3D+SCALE(i%2C+0%2C+sine_table_size%2C+0.0f%2C+PI)%3B%0D%0A++++++++++++sine_table%5Bi%5D+%3D+std%3A%3Asin(angle)%3B%0D%0A++++++++%7D%0D%0A++++%7D%0D%0A%7D%3B%0D%0Astatic+sine_initter+sinner%3B%0D%0A%0D%0Afloat+RageFastSin(float+angle)+%7B%0D%0A++++if+(angle+%3D%3D+0)+%7B+return+0%3B+%7D%0D%0A++++float+index+%3D+angle+*+sine_table_index_mult%3B%0D%0A++++int+first_index+%3D+static_cast%3Cint%3E(index)%3B%0D%0A++++int+second_index+%3D+(first_index+%2B+1)+%25+sine_index_mod%3B%0D%0A++++float+remainder+%3D+index+-+first_index%3B%0D%0A++++first_index+%25%3D+sine_index_mod%3B%0D%0A++++float+first+%3D+0.0f%3B%0D%0A++++float+second+%3D+0.0f%3B%0D%0A%23define+SET_SAMPLE(sample)+%5C%0D%0A++++if(sample%23%23_index+%3E%3D+static_cast%3Cint%3E(sine_table_size))+%5C%0D%0A++++%7B+%5C%0D%0A++++++++sample+%3D+-sine_table%5Bsample%23%23_index+-+sine_table_size%5D%3B+%5C%0D%0A++++%7D+%5C%0D%0A++++else+%5C%0D%0A++++%7B+%5C%0D%0A++++++++sample+%3D+sine_table%5Bsample%23%23_index%5D%3B+%5C%0D%0A++++%7D%0D%0A++++SET_SAMPLE(first)%3B%0D%0A++++SET_SAMPLE(second)%3B%0D%0A%23undef+SET_SAMPLE%0D%0A++++return+lerp(remainder%2C+first%2C+second)%3B%0D%0A%7D%0D%0A%0D%0Aint+main()+%7B%0D%0A++++%0D%0A++++std%3A%3Acout+%3C%3C+%22Hello%2C+World!%22+%3C%3C+std%3A%3Aendl%3B%0D%0A%0D%0A++++%2F%2F+Example+angle+to+test%0D%0A++++float+angle+%3D+0.5f%3B%0D%0A%0D%0A++++%2F%2F+Start+measuring+time%0D%0A++++auto+start+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A%0D%0A++++%2F%2F+Call+the+function%0D%0A++++float+result+%3D+RageFastSin(angle)%3B%0D%0A%0D%0A++++%2F%2F+Stop+measuring+time%0D%0A++++auto+stop+%3D+std%3A%3Achrono%3A%3Ahigh_resolution_clock%3A%3Anow()%3B%0D%0A%0D%0A++++%2F%2F+Calculate+the+duration%0D%0A++++auto+duration+%3D+std%3A%3Achrono%3A%3Aduration_cast%3Cstd%3A%3Achrono%3A%3Amicroseconds%3E(stop+-+start)%3B%0D%0A%0D%0A++++%2F%2F+Print+the+result+and+the+time+taken%0D%0A++++std%3A%3Acout+%3C%3C+%22RageFastSin(%22+%3C%3C+angle+%3C%3C+%22)+%3D+%22+%3C%3C+result+%3C%3C+std%3A%3Aendl%3B%0D%0A++++std%3A%3Acout+%3C%3C+%22Time+taken%3A+%22+%3C%3C+duration.count()+%3C%3C+%22+microseconds%22+%3C%3C+std%3A%3Aendl%3B%0D%0A%0D%0A++++return+0%3B%0D%0A%7D%0D%0A):
```
std::sin(x + 0.5f * PI) = 0.877583
Time taken: 20 microseconds
std::cos(x) = 0.877583
Time taken: 1 microseconds
```

I didn't provide Godbolt links because it would show "0 microseconds" for a lot of these test cases, though I'm sure that's fixable, the main thing was I wanted to try a few different online and offline compilers to ensure the standard library functions are faster.

teejusb commented on src/ArrowEffects.cpp:
```diff
@@ -583,7 +587,8 @@ float ArrowEffects::GetYOffset( const PlayerState* pPlayerState, int iCol, float
 	if( fAccels[PlayerOptions::ACCEL_EXPAND] != 0 )
 	{
-		float fExpandMultiplier = SCALE( RageFastCos(data.m_fExpandSeconds*EXPAND_MULTIPLIER_FREQUENCY*(fAccels[PlayerOptions::ACCEL_EXPAND_PERIOD]+1)),
+		// TODO: Don't index by PlayerNumber.
```
I dont see us indexing my player number

teejusb commented on src/ArrowEffects.cpp:
```diff
@@ -736,20 +741,10 @@ float ArrowEffects::GetXPos( const PlayerState* pPlayerState, int iColNum, float
-					, curr_options->m_bCosecant) * ARROW_SIZE*DRUNK_ARROW_MAGNITUDE );
-
+		fPixelOffsetFromCenter += fEffects[PlayerOptions::EFFECT_DRUNK] * 
+			( std::cos( RageTimer::GetTimeSinceStartFast() + iColNum*DRUNK_COLUMN_FREQUENCY
```
Why is this so different? Keep stuff the same and only do what the PR says it's doing.

teejusb commented on src/ArrowEffects.cpp:
```diff
@@ -1204,13 +1199,10 @@ float ArrowEffects::GetZPos( const PlayerState* pPlayerState, int iCol, float fY
-			CalculateBumpyAngle(fYOffset,
-			fEffects[PlayerOptions::EFFECT_BUMPY_OFFSET],
-			fEffects[PlayerOptions::EFFECT_BUMPY_PERIOD]) );
+		fZPos += fEffects[PlayerOptions::EFFECT_BUMPY] * 40*std::sin( fYOffset/16.0f );
```
Same here

sukibaby commented:
Oh no, it looks like I mixed in some of the really old code from that ancient commit in here 😱  Gonna get that back to how it should be

sukibaby committed:
Use stdlib for trigonometric functions

Profiling showed this performed significantly faster than the RageFast_ implementations.

sukibaby commented:
Resolved the mentioned issues (also altered my workflow to prevent this from happening again).

teejusb commented on src/RageDisplay.cpp:
```diff
@@ -51,7 +51,7 @@ static std::vector<Centering> g_CenteringStack( 1, Centering(0, 0, 0, 0) );
 
 RageDisplay*		DISPLAY	= nullptr; // global and accessible from anywhere in our program
 
-Preference<bool>  LOG_FPS( "LogFPS", false );
```
This is being reverted (accident?)

sukibaby replied:
Yes, thanks 😓

sukibaby commented:
I feel a little weird about taking out the RageFastCos and *2 from this line in this PR instead of in the "Standalone mt19937 RNG function" PR, but it's the only way to prevent a merge conflict.

sukibaby commented:
### There is no longer a merge conflict between this and the mt19937 PR

teejusb commented on src/NoteField.cpp:
```diff
@@ -868,7 +866,7 @@ void NoteField::DrawPrimitives()
 
 		const TimingData &timing = *pTiming;
-		const RageColor text_glow= RageColor(1,1,1,RageFastCos(RageTimer::GetTimeSinceStartFast()*2)/2+0.5f);
+		const RageColor text_glow= RageColor(1,1,1,(RageTimer::GetTimeSinceStartFast()));
```
Let's keep the original math here and below

sukibaby replied:
I was literally pushing this back here as you commented it!
